Documentation/teaching/lectures/interrupts.rst

```diff
@@ -0,0 +1,558 @@
+   * use IRET
+
+     * pops CS, EIP, EFLAGS
+     * if privilege level changed returns to the old stack and old privilege level
+
+Nested interrupts and exceptions
+--------------------------------
+
+An interrupt handler may preempt both other interrupt handlers and exception handlers.
+On the other hand, an exception handler never preempts an interrupt handler.
+
+.. slide:: Interrupt/Exception nesting
+   :inline-contents: True
+   :level: 2
+
+   .. ditaa::
+      
+                     +                                                 ^
+                     |                                                 |
+                     |                                                 |
+          User Mode  | IRQi                                            |
+                     |                                                 |
+                     |                                                 |
+                  +-------------------------------------------------------+
+                     |                                             iret|
+                     |                                                 |
+        Kernel Mode  v-------+      ^-------+                 ^--------+
+                             |      |       |                 |
+                         IRQj|  iret|   IRQk|             iret|
+                             |      |       |                 |
+                             v------+       v-----+     ^-----+
+                                                  |     |
+                                             IRQn | iret|
+                                                  v-----+
+
+.. slide:: Interrupt context
+   :inline-contents: True
+   :level: 2
+
+    * runs as a result of an IRQ (not of an exception)
+    * there is no 'process' context associated
+    * cannot put the current context to sleep
+      
+      * no sleep
+      * no schedule
+      * no user memory access
+
+.. slide:: Interrupt handling
+   :inline-contents: True
+   :level: 2
+
+   .. ditaa::
+
+
+              phase 1
+        +----------------+
+        |    critical    |               phase 2
+        +----------------+         +-----------------+
+        |                |         |    immediate    |               phase 3
+        | - IRQ disabled |         +-----------------+          +----------------+
+        | - ACK IRQ      +-----+   |                 |          |   deferred     |
+        |                |     +---> - IRQ disabled  |          +----------------+
+        +----------------+         | - device handler|          |                |
+                                   |                 +-----+    | - IRQ enabled  |
+                                   +-----------------+     +----> - execute later|
+                                                                |                |
+                                                                +----------------+
```

Comment:
I think it is worth adding a slide about how Linux specifics about handling interrupts: common interrupt entry, shared handlers, etc.

Reply:
Definitely, but I haven't understood it well enough.